### PR TITLE
Add `@protected` to public `State` method overrides

### PR DESCRIPTION
**Pull request by nate-thegrate**
I don't want to see these methods in my autofill suggestions.

<br>

navigator dispose

### CODE REVIEW

**Comment by nate-thegrate** [comment]
Marking as a draft for now, since this didn't qualify for a test exemption.

**Comment by nate-thegrate** [comment]
I imagine the team's overall preference would be to move forward with this PR, but https://github.com/flutter/flutter/pull/157874 is also a fun option!

**Review by victorsanni — COMMENTED** [review]
lgtm with the caveat that I don't have the domain knowledge to understand the implications of the changes in this PR.

**Review by justinmc — COMMENTED** [review]
I've requested Google tests on this PR. That will be a good indicator on whether or not people are referencing these methods in their apps. It's encouraging that the customer tests are already passing at least.

This all sounds like a good idea to me, but I'm worried it's risky, so ideally we can get some other people to look at this too (maybe @goderbauer when he's back?). I wonder why these methods weren't marked `@protected` in the first place, just an oversight? I see that they are `@protected` in their base class, e.g. [State.initState](https://main-api.flutter.dev/flutter/widgets/State/initState.html) is protected.

Your [design doc](https://docs.google.com/document/d/15Rx894OxW1w2Lcgj-ObDMDyLokDB-zpZluX4yUBqbL8/edit?usp=sharing) and https://github.com/flutter/flutter/pull/157874 make a good case for the extension solution, and we've been up for making exceptions to the styleguide lately, but doing that all over the framework would be a huge departure from the intent of the styleguide. The solution in this PR is probably better.

tl;dr: LGTM but we should get more people to look at this.

**Comment by nate-thegrate** [comment]
Thanks very much @justinmc! Hope you don't mind if I argue a couple of points :)

<br>

> I'm worried it's risky

If, for some reason, someone has `Navigator.of(context).dispose()` in their codebase, it would still compile & run as before: the only difference is that now it will trigger a linter rule, which I believe is a very good thing.

<br>

> we've been up for making exceptions to the styleguide lately, but doing that all over the framework would be a huge departure from the intent of the styleguide

I don't believe the other PR violates any style guidelines, since "extension methods" and "extension types" are separate concepts.

- https://dart.dev/language/extension-methods
- https://dart.dev/language/extension-types

But as far as "the intent of the styleguide", I can see how just having the public class without the extension type wrapper would be more readable for people debugging their own projects 👍

**Comment by nate-thegrate** [comment]
https://github.com/flutter/flutter/pull/157313/checks?check_run_id=32405077901

> ## Google Testing
> Timed out

will try re-running!

**Comment by justinmc** [comment]
Looks like the Google tests are passing which is really encouraging!

I see your point about `extension type` vs. extension methods and it's a lot better than I was thinking. However, I think there is still some level of redirection along the lines of what is called out [in the styleguide](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-extension) in that users that go looking for NavigatorState's build method might be surprised to find that it's not in NavigatorState.

Maybe we can figure out a path forward for this in the triage meeting next week. I could definitely be convinced either way. And I do think we should fix this problem one way or another.

Also it would be nice if we could update the styleguide to explicitly state what our policy is on `extension type`.

File: dev/bots/custom_rules/protect_public_state_subtypes.dart [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  @override
+  String toString() => 'Add "@protected" to public State subtypes';
+}
+
+class _StateSubclassVisitor extends SimpleAstVisitor<void> {
+  final _AstList violationNodes = <(AstNode, String)>[];
+
+  static final Map<InterfaceElement, bool> isStateResultCache = <InterfaceElement, bool>{};
+
+  static bool isState(InterfaceElement element) {
```
Comment (justinmc) [inline dev/bots/custom_rules/protect_public_state_subtypes.dart:57]: Nit: I would make the parameter here and in _isState have the same name.

File: dev/bots/custom_rules/protect_public_state_subtypes.dart [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    return element.allSupertypes.any((InterfaceType interface) => _isState(interface.element));
+  }
+
+  static bool _isState(InterfaceElement interfaceElement) {
+    // Framework naming convention: each State subclass has "State" in its name.
+    if (!interfaceElement.name.contains('State')) {
```
Comment (justinmc) [inline dev/bots/custom_rules/protect_public_state_subtypes.dart:63]: Can you not check the type?

**Review by justinmc — APPROVED** [review]
LGTM :+1: . Thanks for going through all of the back and forth while deciding whether to take this approach or the other one.

I assume your new lint passes on this branch?

I admit I haven't added a custom lint rule before so I'm not super familiar with that code, maybe someone who knows more should take a quick look at that before merging if possible.

File: dev/bots/custom_rules/protect_public_state_subtypes.dart (hunk 2 above)
Comment (nate-thegrate) [reply dev/bots/custom_rules/protect_public_state_subtypes.dart:63]: Sadly no 😞

This is a "metadata object" containing info about the class declaration, I don't believe it ever creates an instance of the class.

File: dev/bots/custom_rules/protect_public_state_subtypes.dart (hunk 1 above)
Comment (nate-thegrate) [reply dev/bots/custom_rules/protect_public_state_subtypes.dart:57]: Good catch, thanks!

File: dev/bots/custom_rules/protect_public_state_subtypes.dart
Comment (matanlurey) [inline dev/bots/custom_rules/protect_public_state_subtypes.dart:1]: I'd personally like to see this tested, so that we know it works and know what happens if we break the rule.

I am not sure there are examples of lint tests today (there are in `package:analyzer`), so I'd take a TODO and a github issue if not.

**Review by matanlurey — CHANGES_REQUESTED** [review]

File: dev/bots/custom_rules/protect_public_state_subtypes.dart [hunk 3]
```diff
@@ -2,6 +2,9 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+// TODO(nate-thegrate): remove this file if @protected changes, or add a test if it doesn't.
+// https://github.com/dart-lang/sdk/issues/57094
```
Comment (nate-thegrate) [inline dev/bots/custom_rules/protect_public_state_subtypes.dart:6]: I think the conversation there is pretty interesting:

- https://github.com/dart-lang/sdk/issues/57094

File: dev/bots/custom_rules/protect_public_state_subtypes.dart [hunk 4]
```diff
@@ -50,34 +51,36 @@ class _ProtectPublicStateSubtypes implements AnalyzeRule {
...
-  final _AstList violationNodes = <(AstNode, String)>[];
+  final List<MethodDeclaration> unprotectedMethods = <MethodDeclaration>[];
 
-  static final Map<InterfaceElement, bool> isStateResultCache = <InterfaceElement, bool>{};
-
-  static bool isState(InterfaceElement element) {
-    return element.allSupertypes.any((InterfaceType interface) => _isState(interface.element));
-  }
+  /// Holds the `State` class [DartType].
+  static DartType? stateType;
```
Comment (nate-thegrate) [inline dev/bots/custom_rules/protect_public_state_subtypes.dart:57]: ### 😌

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -1495,6 +1495,9 @@ abstract class State<T extends StatefulWidget> with Diagnosticable {
     properties.add(ObjectFlagProperty<T>('_widget', _widget, ifNull: 'no widget'));
     properties.add(ObjectFlagProperty<StatefulElement>('_element', _element, ifNull: 'not mounted'));
   }
+
+  // If @protected State methods are added or removed, the analysis rule should be
+  // updated accordingly (dev/bots/custom_rules/protect_public_state_subtypes.dart)
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/framework.dart:1500]: (appended this to the `State` class declaration)

**Review by nate-thegrate — COMMENTED** [review]
Just caching a single `DartType` was the right call 👍

File: dev/bots/custom_rules/protect_public_state_subtypes.dart
Comment (nate-thegrate) [reply dev/bots/custom_rules/protect_public_state_subtypes.dart:1]: https://github.com/flutter/flutter/pull/157313#discussion_r1844738734

File: dev/bots/custom_rules/protect_public_state_subtypes.dart [hunk 6]
```diff
@@ -0,0 +1,103 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply dev/bots/custom_rules/protect_public_state_subtypes.dart:1]: resolving per https://github.com/flutter/flutter/pull/157313#discussion_r1844739512

**Review by matanlurey — APPROVED** [review]

**Comment by nate-thegrate** [comment]
```
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':app:compressDebugAssets'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.CompressAssetsWorkAction
   > Java heap space
```

(assuming this is a flake)
